### PR TITLE
[TE] Create detection onboarding execution pipeline

### PR DESCRIPTION
**Pull request by cyenjung**
Create an in-memory job executor that runs job's linear tasks. This executor satifies the following requirement for detection onboarding jobs:
1. One job could have multiple tasks.
2. Error message of task and job execution could be propagate back to job executor.
3. Task failure could be ignore (i.e., job execution continues) if "abortAtFailure = false" for that task.

Tested with unit tests to ensure the requirements are fullfilled.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardExecutionContext.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.linkedin.thirdeye.anomaly.onboard;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class DetectionOnboardExecutionContext {
```
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardExecutionContext.java]: Please provide a java doc describing the the context here.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardJob.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package com.linkedin.thirdeye.anomaly.onboard;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.configuration.Configuration;
+
+public interface DetectionOnboardJob {
```
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardJob.java]: Should the interface restrict to detection onboard job? Do you have plans to extend this design to all jobs in ThirdEye?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardService.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.linkedin.thirdeye.anomaly.onboard;
+
+import java.util.Map;
+
+public interface DetectionOnboardService {
```
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardService.java]: Will we have multiple detection onboarding services? What is the plan here? Or it is to extend it to the other job services?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardJob.java (hunk 2 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardJob.java]: We could explore the possibility.

In order to extend this temporary executor to TE's generic executor, we need to make sure we can persistent execution status of job and tasks, which shouldn't be too difficult.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardService.java (hunk 3 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardService.java]: I am hoping this class will be extended to an interface for generic job executor.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardExecutionContext.java (hunk 1 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardExecutionContext.java]: Added.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java [hunk 4]
```diff
@@ -133,34 +136,42 @@ public void start() throws Exception {
...
+          detectionOnboardServiceExecutor = new DetectionOnboardServiceExecutor();
+          detectionOnboardServiceExecutor.start();
+          environment.jersey().register(new DetectionOnboardResource(detectionOnboardServiceExecutor));
+        }
       }
 
       @Override
       public void stop() throws Exception {
-        if (config.isWorker()) {
+        if (config.isWorker() && taskDriver != null) {
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java]: the first part is redundant. for cleanup use only "taskDriver != null"

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/BaseDetectionOnboardTask.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import org.apache.commons.lang.StringUtils;
+
+public abstract class BaseDetectionOnboardTask implements DetectionOnboardTask {
+  private final String taskName;
+  protected DetectionOnboardTaskContext taskContext = new DetectionOnboardTaskContext();
+
+  public BaseDetectionOnboardTask(String taskName) {
+    Preconditions.checkNotNull(taskName);
+    Preconditions.checkArgument(StringUtils.isNotBlank(taskName.trim()));
+    this.taskName = taskName.trim();
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/BaseDetectionOnboardTask.java]: i'd just pass through the actual name

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+  private final DetectionOnboardJobContext jobContext;
+  private final List<DetectionOnboardTask> tasks;
+  private final DetectionOnboardJobStatus jobStatus;
+  private final int taskTimeOutSize;
+  private final TimeUnit taskTimeOutUnit;
+
+  public DetectionOnBoardJobRunner(DetectionOnboardJobContext jobContext, List<DetectionOnboardTask> tasks,
+      DetectionOnboardJobStatus jobStatus) {
+    this(jobContext, tasks, jobStatus, 5, TimeUnit.MINUTES);
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java]: might as well poll every minute

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        task.setTaskContext(taskContext);
+        taskStatus.setTaskStatus(TaskConstants.TaskStatus.RUNNING);
+        Future<DetectionOnboardTaskStatus> taskFuture = executorService.submit(new DetectionOnboardTaskRunner(task));
+        // Wait until time out
+        DetectionOnboardTaskStatus returnedTaskStatus = taskFuture.get(taskTimeOutSize, taskTimeOutUnit);
+        taskStatus.setTaskStatus(returnedTaskStatus.getTaskStatus());
+        taskStatus.setMessage(returnedTaskStatus.getMessage());
+      } catch (TimeoutException e) {
+        taskStatus.setTaskStatus(TaskConstants.TaskStatus.TIMEOUT);
+        LOG.warn("Task {} is timed out.", task.getTaskName());
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java]: delete "is"

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    Preconditions.checkNotNull(tasks);
+    Preconditions.checkNotNull(jobStatus);
+
+    for (DetectionOnboardTask task : tasks) {
+      DetectionOnboardTaskStatus taskStatus = new DetectionOnboardTaskStatus(TaskConstants.TaskStatus.WAITING, "");
+      jobStatus.addTaskStatus(taskStatus);
+
+      // Construct Task context and configuration
+      Configuration taskConfig = jobContext.getConfiguration().subset(task.getTaskName());
+      final boolean abortAtFailure = taskConfig.getBoolean("abortAtFailure", true);
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java]: "abortOnFailure"

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardJob.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+   * @return the unique name of this job.
+   */
+  String getName();
+
+  /**
+   * Initialize this job with the given properties.
+   *
+   * @param properties the properties to initialize the job.
+   */
+  void initialize(Map<String, String> properties);
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardJob.java]: this "should" be the constructor. IMO avoid init() methods whenever you can. Pinot just underwent a major cleanup for this

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardResource.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  @GET
+  @Path("/get-status")
+  public String getDetectionOnboardingJobStatus(@QueryParam("jobId") long jobId) {
+    DetectionOnboardJobStatus detectionOnboardingJobStatus =
+        detectionOnboardService.getDetectionOnboardingJobStatus(jobId);
+    // Create StatusNotFound message
+    if (detectionOnboardingJobStatus == null) {
+      detectionOnboardingJobStatus = new DetectionOnboardJobStatus();
+      detectionOnboardingJobStatus.setJobId(jobId);
+      detectionOnboardingJobStatus.setJobStatus(JobConstants.JobStatus.FAILED);
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardResource.java]: best practice add an 'unknown' job state or throw an exception

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardService.java (hunk 3 above)
Comment (apucher) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardService.java]: I wonder if this duplicates existing logic - i.e. could any job submission simply return the id assigned by the database, without keeping track of it in memory? What if there are multiple different frontend servers?

File: thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunnerTest.java [hunk 11]
```diff
@@ -0,0 +1,318 @@
+package com.linkedin.thirdeye.anomaly.onboard;
```
Comment (apucher) [inline thirdeye/thirdeye-pinot/src/test/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunnerTest.java]: +1 for extensive unit testing

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardService.java (hunk 3 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardService.java]: This executor service does not use DB and hence the job id is assigned and tracked in its memory. The next step is to use TE's DB for distributing the tasks. At that time, the Job ID will be the one returned from DB. In addition, we assume that there is only one machine runs detection onboarding jobs.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java (hunk 4 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java]: Good observation.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java (hunk 6 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java]: This is max execution time for a task.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java (hunk 8 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java]: Fixed.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java (hunk 7 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnBoardJobRunner.java]: Fixed.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardJob.java (hunk 9 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardJob.java]: Fixed.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardResource.java (hunk 10 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/DetectionOnboardResource.java]: Nice suggestion.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/BaseDetectionOnboardTask.java (hunk 5 above)
Comment (cyenjung) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/onboard/BaseDetectionOnboardTask.java]: Fixed.

**Pull request merged**